src/server/scripts/Outland/Auchindoun/AuchenaiCrypts/boss_exarch_maladaar.cpp

```diff
@@ -15,305 +15,365 @@
 };
 
-enum ExarchMaladaar
+enum Misc
 {
-    SAY_INTRO                   = 0,
-    SAY_SUMMON                  = 1,
-    SAY_AGGRO                   = 2,
-    SAY_ROAR                    = 3,
-    SAY_SLAY                    = 4,
-    SAY_DEATH                   = 5,
-
-    SPELL_RIBBON_OF_SOULS       = 32422,
-    SPELL_SOUL_SCREAM           = 32421,
-    SPELL_STOLEN_SOUL           = 32346,
-    SPELL_STOLEN_SOUL_VISUAL    = 32395,
-    SPELL_SUMMON_AVATAR         = 32424,
-
-    ENTRY_STOLEN_SOUL           = 18441
+    NPC_DORE                    = 19412,
+
+    MODEL_UNDEAD_MALE           = 1027,    // guessed
+    MODEL_UNDEAD_FEMALE         = 1029,    // guessed
+    MODEL_NIGHTELF_MALE         = 2572,    // guessed
+    MODEL_NIGHTELF_FEMALE       = 2575,    // guessed
+    MODEL_ORC_MALE              = 2576,    // guessed
+    MODEL_ORC_FEMALE            = 2577,    // guessed
+    MODEL_TAUREN_MALE           = 2578,    // sniff
+    MODEL_TAUREN_FEMALE         = 2579,    // guessed
+    MODEL_GNOME_MALE            = 2581,    // guessed
+    MODEL_GNOME_FEMALE          = 2590,    // guessed
+    MODEL_HUMAN_MALE            = 2582,    // sniff
+    MODEL_HUMAN_FEMALE          = 2583,    // sniff
+    MODEL_DWARF_MALE            = 2584,    // sniff
+    MODEL_DWARF_FEMALE          = 2585,    // guessed
+    MODEL_TROLL_MALE            = 2588,    // guessed
+    MODEL_TROLL_FEMALE          = 2589,    // guessed
+    MODEL_BLOODELF_MALE         = 17267,   // completely guessed
+    MODEL_BLOODELF_FEMALE       = 17268,   // completely guessed
+    MODEL_DRAENEI_MALE          = 16721,   // completely guessed
+    MODEL_DRAENEI_FEMALE        = 17004    // completely guessed
 };
 
-class boss_exarch_maladaar : public CreatureScript
+Position const DoreSpawnPos = { -4.40722f, -387.277f, 40.6294f, 6.26573f };
+
+struct boss_exarch_maladaar : public BossAI
 {
-public:
-    boss_exarch_maladaar() : CreatureScript("boss_exarch_maladaar") { }
+    boss_exarch_maladaar(Creature* creature) : BossAI(creature, DATA_EXARCH_MALADAAR), _avatarSummoned(false) { }
 
-    CreatureAI* GetAI(Creature* creature) const override
+    void Reset() override
     {
-        return GetAuchenaiCryptsAI<boss_exarch_maladaarAI>(creature);
+        _Reset();
+        _avatarSummoned = false;
     }
 
-    struct boss_exarch_maladaarAI : public ScriptedAI
+    void JustEngagedWith(Unit* who) override
     {
-        boss_exarch_maladaarAI(Creature* creature) : ScriptedAI(creature)
+        BossAI::JustEngagedWith(who);
+        Talk(SAY_AGGRO);
+        events.ScheduleEvent(EVENT_SOUL_SCREAM, RAND(10s, 15s, 20s, 25s));
+        events.ScheduleEvent(EVENT_RIBBON_OF_SOULS, 1s, 4s);
+        events.ScheduleEvent(EVENT_STOLEN_SOUL, RAND(10s, 15s, 20s, 25s));
+    }
+
+    void DamageTaken(Unit* /*attacker*/, uint32& /*damage*/, DamageEffectType /*damageType*/, SpellInfo const* /*spellInfo = nullptr*/) override
+    {
+        if (!_avatarSummoned && !HealthAbovePct(25))
```

Comment:
me->HealthBelowPctDamaged(25, damage); to ensure that the event is being scheduled right when the health falls below 25% pct.

Reply:
I was thinking about that because currently it's indeed scheduled in a weird way, can't even figure out why eventually it's always scheduled
I'll try later, thanks

Just in case here it's in separate event because of https://github.com/TrinityCore/TrinityCore/pull/27189#issuecomment-954938274

Reply:
that's fine, just use my suggested helper and you will get more consistent results. The problem with HealthAbovePct is that it returns false AFTER taking damage because   the DamageTaken hook is being called before the damage was applied on the health. HealthBelowPctDamage instead uses the incoming damage and checks it against the current health to handle exactly that problem.